grabbit/src/test/groovy/com/twcable/grabbit/server/batch/steps/jcrnodes/JcrNodesProcessorSpec.groovy

```diff
@@ -69,4 +77,62 @@ class JcrNodesProcessorSpec extends Specification {
     }
 
+    def "Can marshall a JCR Node with mandatory child node to a Protobuf Message"() {
+        given:
+        // Setup parent node as nt:file which has mandatory child node
+        JcrNode hierarchyNode = Mock()
+        hierarchyNode.path >> "default.groovy"
+        JcrProperty primaryProperty = Mock()
+        primaryProperty.type >> 1
+        primaryProperty.name >> JcrConstants.JCR_PRIMARYTYPE
+        primaryProperty.value >> new ValueImpl(JcrConstants.NT_FILE)
+        PropertyIterator propertyIterator = new PropertyIteratorAdapter([primaryProperty].iterator())
+        hierarchyNode.properties >> propertyIterator
+
+        NodeDefinition nodeDefinition = Mock()
+        hierarchyNode.definition >> nodeDefinition
+
+        NodeType nodeType = Mock()
+        NodeDefinition childNodeDefinition = Mock()
+        childNodeDefinition.mandatory >> true
+        nodeType.childNodeDefinitions >> [childNodeDefinition]
+        hierarchyNode.primaryNodeType >> nodeType
+
+        // Setup child node "default.groovy/jcr:content", which is mandatory child node and a property type as nt:resource
+        JcrNode childMandatoryNode = Mock()
+        childMandatoryNode.path >> "default.groovy/jcr:content"
+        JcrProperty childPrimaryProperty = Mock()
+        childPrimaryProperty.type >> 1
+        childPrimaryProperty.name >> JcrConstants.JCR_PRIMARYTYPE
+        childPrimaryProperty.value >> new ValueImpl(JcrConstants.NT_RESOURCE)
+        PropertyIterator childPropertyIterator = new PropertyIteratorAdapter([childPrimaryProperty].iterator())
+        childMandatoryNode.properties >> childPropertyIterator
+        childMandatoryNode.definition >> childNodeDefinition
+
+        childMandatoryNode.parent >> hierarchyNode
+
+        NodeType childNodeType = Mock()
+        childNodeType.childNodeDefinitions >> [Mock(NodeDefinition)]
+        childMandatoryNode.primaryNodeType >> childNodeType
+        NodeIterator nodeIterator = new NodeIteratorAdapter([childMandatoryNode].iterator())
+        hierarchyNode.getNodes() >> nodeIterator
+
+        when:
+        JcrNodesProcessor jcrNodesProcessor = new JcrNodesProcessor()
+        NodeProtos.Node nodeProto = jcrNodesProcessor.process(hierarchyNode)
+
+        then:
+        nodeProto != null
+        nodeProto.name == "default.groovy"
+        // Verified parent node as nt:file which has mandatory child node
+        nodeProto.properties.propertyList.any {
+            it.name == JcrConstants.JCR_PRIMARYTYPE && it.value.stringValue == JcrConstants.NT_FILE
+        }
+        // Verified child node "default.groovy/jcr:content", which is mandatory child node and a property type as nt:resource
+        nodeProto.mandatoryChildNodeList.any {
+            it.name == "default.groovy/jcr:content" && it.properties.propertyList.any {
+                it.name == JcrConstants.JCR_PRIMARYTYPE && it.value.stringValue == JcrConstants.NT_RESOURCE
+            }
+        }
+    }
```

Comment:
Rather than that, which is awfully hard to follow the setup, a cleaner and more Spock-like way would be
```groovy
def "Can marshall a JCR Node with mandatory child node to a Protobuf Message"() {
    given:
    def mandatoryChildNodeDefinition = Mock(NodeDefinition) {
        isMandatory() >> true
    }

    // Setup parent node as nt:file which has mandatory child node
    def hierarchyNode = Mock(JcrNode) {
        getPath() >> "default.groovy"
        getDefinition() >> Mock(NodeDefinition)
        getProperties() >> new PropertyIteratorAdapter([Mock(JcrProperty) {
            getType() >> STRING
            getName() >> JcrConstants.JCR_PRIMARYTYPE
            getValue() >> new ValueImpl(JcrConstants.NT_FILE)
        }].iterator())
        getPrimaryNodeType() >> Mock(NodeType) {
            getChildNodeDefinitions() >> [mandatoryChildNodeDefinition]
        }
    }

    // Setup child node "default.groovy/jcr:content", which is mandatory child node and a property type as nt:resource
    def childMandatoryNode = Mock(JcrNode) {
        getPath() >> "default.groovy/jcr:content"
        getDefinition() >> mandatoryChildNodeDefinition
        getProperties() >> new PropertyIteratorAdapter([Mock(JcrProperty) {
            getType() >> STRING
            getName() >> JcrConstants.JCR_PRIMARYTYPE
            getValue() >> new ValueImpl(JcrConstants.NT_RESOURCE)
        }].iterator())
        getParent() >> hierarchyNode
        getPrimaryNodeType() >> Mock(NodeType) {
            getChildNodeDefinitions() >> [Mock(NodeDefinition)]
        }
    }

    hierarchyNode.getNodes() >> new NodeIteratorAdapter([childMandatoryNode].iterator())

    when:
    NodeProtos.Node nodeProto = new JcrNodesProcessor().process(hierarchyNode)

    then:
    nodeProto.name == "default.groovy"
    nodeProto.properties.propertyList.first().value.stringValue == JcrConstants.NT_FILE
    nodeProto.mandatoryChildNodeList.size() == 1
    nodeProto.mandatoryChildNodeList.first().name == "default.groovy/jcr:content"
}
```
The `then:` section is also greatly simplified to only test what we care about.

Or the test could be simplified even more simply with
```groovy
def "Can marshall a JCR Node with mandatory child node to a Protobuf Message "() {
    given:
    def hierarchyNode = createNode("default.groovy", false, JcrConstants.NT_FILE,
            [createNode("default.groovy/jcr:content", true, JcrConstants.NT_RESOURCE)])

    when:
    NodeProtos.Node nodeProto = new JcrNodesProcessor().process(hierarchyNode)

    then:
    nodeProto.name == "default.groovy"
    nodeProto.properties.propertyList.first().value.stringValue == JcrConstants.NT_FILE
    nodeProto.mandatoryChildNodeList.size() == 1
    nodeProto.mandatoryChildNodeList.first().name == "default.groovy/jcr:content"
}

NodeDefinition mandatoryNodeDefinition() {
    return Mock(NodeDefinition) {
        isMandatory() >> true
    }
}


NodeDefinition nonMandatoryNodeDefinition() {
    return Mock(NodeDefinition) {
        isMandatory() >> false
    }
}


JcrProperty primaryTypeProperty(String propertyValue) {
    return Mock(JcrProperty) {
        getType() >> STRING
        getName() >> JcrConstants.JCR_PRIMARYTYPE
        getValue() >> new ValueImpl(propertyValue)
    }
}


static PropertyIterator propertyIterator(JcrProperty... properties) {
    new PropertyIteratorAdapter(properties.iterator())
}


JcrNode createNode(String path, boolean isMandatory, String primaryType, Collection<JcrNode> children = []) {
    def nodeDefinition = isMandatory ? mandatoryNodeDefinition() : nonMandatoryNodeDefinition()

    def childDefinitions = children.collect {it.getDefinition()} as List<NodeDefinition>

    def node = Mock(JcrNode) {
        getPath() >> path
        getDefinition() >> nodeDefinition
        getProperties() >> propertyIterator(primaryTypeProperty(primaryType))
        getPrimaryNodeType() >> Mock(NodeType) {
            getChildNodeDefinitions() >> childDefinitions.toArray()
        }
    }

    children.each { JcrNode child ->
        child.getParent() >> node
    }
    node.getNodes() >> new NodeIteratorAdapter(children.iterator())

    return node
}
```
The advantage here is that the feature itself becomes trivial to not only create the node structure, but also to understand it.